Userland/Libraries/LibAudio/FlacLoader.cpp

```diff
@@ -0,0 +1,427 @@
+/*
+ * Copyright (c) 2021, kleines Filmröllchen <malu.bertsch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "FlacLoader.h"
+#include <AK/BitStream.h>
+#include <AK/Debug.h>
+#include <AK/FlyString.h>
+#include <AK/Stream.h>
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/File.h>
+#include <LibCore/FileStream.h>
+#include <LibCore/IODeviceStreamReader.h>
+#include <LibCrypto/Checksum/CRC8.h>
+#include <LibM/math.h>
+
+namespace Audio {
+
+FlacLoaderPlugin::FlacLoaderPlugin(const StringView& path)
+    : m_file(Core::File::construct(path))
+{
+    if (!m_file->open(Core::IODevice::ReadOnly)) {
+        m_error_string = String::formatted("Can't open file: {}", m_file->error_string());
+        return;
+    }
+
+    m_valid = parse_header();
+    if (!m_valid)
+        return;
+
+    m_stream = make<FlacInputStream>(Core::InputFileStream(*m_file));
+    reset();
+
+    m_resampler = make<ResampleHelper>(m_sample_rate, 44100);
+}
+
+FlacLoaderPlugin::FlacLoaderPlugin(const ByteBuffer& buffer)
+{
+    m_stream = make<FlacInputStream>(InputMemoryStream(buffer));
+    if (!m_stream) {
+        m_error_string = String::formatted("Can't open memory stream");
+        return;
+    }
+
+    m_valid = parse_header();
+    if (!m_valid)
+        return;
+    reset();
+
+    m_resampler = make<ResampleHelper>(m_sample_rate, 44100);
+}
+
+bool FlacLoaderPlugin::sniff()
+{
+    return m_valid;
+}
+
+bool FlacLoaderPlugin::parse_header()
+{
+    Optional<Core::InputFileStream> fis;
+    bool ok = true;
+
+    InputBitStream bit_input = [&]() -> InputBitStream {
+        if (m_file) {
+            fis = Core::InputFileStream(*m_file);
+            return InputBitStream(*fis);
+        }
+        return InputBitStream(m_stream->memory);
+    }();
+
+#define CHECK_OK(msg)                                                      \
```

Comment:
these macros look rather jenky. macros are rather jenky in general though. is there a better way to do this? maybe a lambda?

Reply:
That is the WavLoader design, but yes, I'll find a better way.

Reply:
After looking at alternatives yesterday, it looks like this may actually be quite difficult to change. I'll need to conditionally return from the function, which is not possible in a lambda. So instead I would need to put in a bunch of cluttering `if (check) return` statements.